_posts/2019-03-XX-5-terms-every-hypervisor-developer-should-know.md

```diff
@@ -0,0 +1,292 @@
+environment* and everything seen by one core (general purpose registers, MSRs,
+APIC, memory etc.) is a *virtual machine*. This definition of VM is more or less
+consistent with how they are managed under the hood. I'll try to stick to these
+names, but bear with me if I make a mistake at some point :)
+
+All VMs work in VMX non-root operation, they are called `guest` in SDM as well
+as by everyone else, but...
+
+#### Host-VM
+
+There is one special VM that was created as first VM, it is a virtualized
+version of the operating system or firmware that hypervisor was started with.
+Note that it is sometimes called as a host (because it was there before
+hypervisor) and sometimes as a guest (because it is a VM after all), depending
+on context.
+
+Handlers for this VM, with few exceptions, pass through all accesses to the
+hardware in order to make continued operation of the OS/FW possible.
+
+In theory, this VM isn't necessary, but it is convenient to leave it. Host-VM
+provides many useful parts, such as memory management, ACPI tables, firmware for
+peripheral devices etc. Hypervisor would have to do all that and much more by
+itself if it wasn't for this VM.
+
+## Transitions
+
+There are only two types of transition: VM-exit and VM-entry. As you can see,
+they are named relative to the VM, as this is where the CPU should spend most of
+the time. VM-exit is a transition from VM to VMM, or non-root to root operation
+(these are synonymous); VM-entry - the other way around.
+
+Multiple checks are performed both on entries and exits. When an error occurred
+during VM entry it is possible to just not get into VM or return as soon as
+error on VM side happens. Error on VM exit is worse - they happen only when
+there is no valid VMM to get back to, which leads to VMX abort, after which
+the processor is put into a shutdown state.
+
+Old state is saved and a new one is loaded from VMCS (virtual machine control
+structure, described later) or structures that VMCS points to. There is a field
+for MSRs table, so VMM can fill in the ones that it intends to change and they
+will be saved/restored as a part of the transition.
+
+#### VM entries
+
+VM entries happen as a result of VMLAUNCH or VMRESUME. Steps are done in order:
+
+1. Basic checks are performed to ensure that VM entry can commence (valid VMM
+   state, valid VMCS).
+2. The control and host-state areas of the VMCS are checked to ensure that they
+   are proper for supporting VMX non-root operation and that the VMCS is
+   correctly configured to support the next VM exit.
+3. The guest-state area of the VMCS is checked to ensure that, after the VM
+   entry completes, the state of the logical processor is legal.
+4. MSRs are loaded from the VM-entry MSR-load area.
+5. If VMLAUNCH is being executed, the launch state of the VMCS is set to
+   "launched".
+6. An event may be injected in the guest context.
+
+#### VM exits
```

Comment:
@krystian-hebel some diagram in draw.io would be great - nothing fancy, but something that present idea

Reply:
@pietrushnic I'm not sure if this is good idea, there is too much text to make it readable

Reply:
@krystian-hebel as you wish, that was just an idea.